src/copy.jl

```diff
@@ -0,0 +1,128 @@
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+    copy_extension_data(data, new_model::AbstractModel, model::AbstractModel)
+
+Return a copy of the extension data `data` of the model `model` to the extension
+data of the new model `new_model`. A method should be added for any JuMP
+extension storing data in the `ext` field.
+"""
+function copy_extension_data end
+
+"""
+    copy_variablewise_constraints(dest::Dict{MOIVAR,
+                                             MOICON{MOI.SingleVariable, S}},
+                                  src::Dict{MOIVAR,
+                                            MOICON{MOI.SingleVariable, S}},
+                                  index_map) where S
+
+Copy the variablewise constraint indices of `src` into `dest` mapping variable
+and constraint indices using `index_map`.
+"""
+function copy_variablewise_constraints(dest::Dict{MOIVAR,
+                                                  MOICON{MOI.SingleVariable, S}},
+                                       src::Dict{MOIVAR,
+                                                 MOICON{MOI.SingleVariable, S}},
+                                       index_map) where S
+    for (variable_index, constraint_index) in src
+        dest[index_map[variable_index]] = index_map[constraint_index]
+    end
+end
+
+"""
+    ReferenceMap
+
+Mapping between variable and constraint reference of a model and its copy. The
+reference of the copied model can be obtained by indexing the map with the
+reference of the corresponding reference of the original model.
+"""
+struct ReferenceMap
+    model::Model
+    index_map::MOIU.IndexMap
+end
+function Base.getindex(reference_map::ReferenceMap, vref::VariableRef)
+    return VariableRef(reference_map.model,
+                       reference_map.index_map[index(vref)])
+end
+function Base.getindex(reference_map::ReferenceMap, cref::ConstraintRef)
+    return ConstraintRef(reference_map.model,
+                         reference_map.index_map[index(cref)],
+                         cref.shape)
+end
+
+"""
+    copy(model::Model)
+
+Return a copy of the model `model` and a [`ReferenceMap`](@ref) that can be used
+to obtain the variable and constraint reference of the new model corresponding
+to a given `model`'s reference.
+
+## Note
+
+Model copy is not supported in Direct mode, i.e. when a model is constructed
+using the [`direct_model`](@ref) constructor instead of the [`Model`](@ref)
+constructor.
+
+## Examples
+
+In the following example, a model `model` is constructed with a variable `x` and
+a constraint `cref`. It is then copied into a model `new_model` with the new
+references assigned to `x_new` and `cref_new`.
+```julia
+model = Model()
+@variable(model, x)
+@constraint(model, cref, x == 2)
+
+new_model, reference_map = JuMP.copy(model)
```

Comment:
What do you think about not returning a `reference_map` and instead telling users to look up the variables by name or symbol? It would be nice to be able to match the signature of `Base.copy` to avoid prefixing all of its uses.  Or maybe have `Base.copy` return only the model and `JuMP.copy_model` return both the model and the reference_map?

Reply:
So if the user use `Base.copy` then he is forced to use the `objdict` and if he uses `JuMP.copy_model` then he can use both the reference map and the `objdict`. Is that what you suggest ?

Reply:
Yes (except you can use both lookup by symbol and lookup by name) in either case.